### PR TITLE
Link Storage and Availability Zone in VPC

### PR DESCRIPTION
**Pull request by blancett**

### CODE REVIEW

File: app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb [hunk 1]
```diff
@@ -200,14 +217,27 @@ def floating_ips
...
-        :size          => vol[:capacity]&.gigabytes
+        :ems_ref              => vol[:id],
+        :name                 => vol[:name],
+        :status               => vol[:status],
+        :creation_time        => vol[:created_at],
+        :description          => 'IBM Cloud Block-Storage Volume',
+        :volume_type          => vol[:type],
+        :size                 => vol[:capacity]&.gigabytes,
+        :bootable             => bootable,
+        :availability_zone_id => persister.availability_zones.lazy_find(az_name)
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb:240]: This should be `:availability_zone => persister.availability_zones.lazy_find(az_name)`

File: app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb [hunk 2]
```diff
@@ -200,14 +217,25 @@ def floating_ips
 
   def volumes
     collector.volumes.each do |vol|
+      az_name = vol&.dig(:zone, :name)
+      attachments = vol&.dig(:volume_attachments)
+      bootable = attachments.any?
+      if bootable
+        attachments.each do |vol_attach|
+          bootable = vol_attach[:type] == "boot"
+          break if bootable
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb:226]: This looks like it can be `bootable_volume = attachments.to_a.detect { |vol| vol[:type] == "boot" }`

File: app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb [hunk 3]
```diff
@@ -93,6 +94,22 @@ def instance_hardware(persister_instance, instance)
...
+
+  def instance_storage(persister_hardware, instance)
+    instance[:volume_attachments].each do |vol_attach|
+      vol = collector.volume(vol_attach&.dig(:volume, :id))
+      persister.disks.build(
+        :hardware        => persister_hardware,
+        :device_name     => vol[:name],
+        :device_type     => vol[:type],
+        :controller_type => "ibm",
+        :backing         => persister.cloud_volumes.find(vol[:id]),
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb:108]: Best to use lazy_find instead of find, then the order doesn't matter

File: app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb [hunk 4]
```diff
@@ -18,8 +18,8 @@ def parse
     auth_key_pairs
     flavors
     images
-    instances
     volumes
+    instances
   end
```
Comment (chessbyte) [inline app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb:23]: why this change in the order?

Comment (agrare) [reply app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb:23]: Probably because https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/76#discussion_r501380989 was using `.find()` so the volume had to be in the collection already.

With `.lazy_find()` the order shouldn't matter

Comment (blancett) [reply app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb:23]: @chessbyte the instance needed to find its volume allocation and needed volumes built to find the size. As @agrare pointing out the order mattered because was using find vs lazy_find. I have switched to lazy_find and setting the order back. Thanks for the catch

File: app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb (hunk 2 above)
Comment (blancett) [reply app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb:225]: nice - much cleaner - will switch

File: app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb (hunk 1 above)
Comment (blancett) [reply app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb]: fixed

File: app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb (hunk 2 above)
Comment (blancett) [reply app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb]: fixed

File: app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb (hunk 4 above)
Comment (blancett) [reply app/models/manageiq/providers/ibm_cloud/inventory/parser/vpc.rb]: fixed

File: spec/vcr_cassettes/manageiq/providers/ibm_cloud/vpc/cloud_manager/refresher.yml [hunk 5]
```diff
@@ -32,18 +32,20 @@ http_interactions:
...
...0tshNDXm4XHizJBJY-h9_I5PHUN7gSWUhzmeDvJO3tOi6x6ePJg9rcVZw6omJ-sOuCwi1VvCkplI01A_5fXBkbEVQG6i3AR5pSXURDuIgUgaNgTkvq6nW1-QmX2fUwB7HAMmG-e3IyYbTVnNRlIqhy8_-XpLEDb4wwHp9Ran5gL0_-jEOj3LCUI1zOGHhz9IgeV-jaQVAWp5bZ9vYrroEOMTcjKqYmo_1JGZtyhk9e2SSB5c89-i4ggoMjM3cXOhJgR7wQNHVCA8xCSzAKrcV5Qr86Tqk7GHiCPLslS8eT_6xW_MyU8ipin78drrdNUXd4vg-ir3pnLPwXwxC_Up0f07sIUgaBr-GFoQaBmCl3s21NPE-4AWC32esdZiMNWmXOmLfrwgBCWalQZ6i-1Sx3eKrmAtWVtPcYFYJ2TtaVwG83znBX0HidAw8oZv4iDhC-BwxGOmTS8lIsSD_6Fw2H7TIqN7Z0ie0vkGijRifOHAV0RyFOE4ZPSP5hay_UHZ4qh8O-JtTNmxPuOqXy0ypgu4T6f4RMn5Exd4eusIix79r82FUS0Ejo5YKKND6_G54pybyK228pdrso9_JcEYzMkUjrA2I5lvsOb2EWk_r16MOnvVN18mH_8RatrBlmBB3HZBivHh9S969JTAWmVTcgvcghKLq_yn9rjisGmrhyqH_tUrgRnrUrfXccuerLkkg5Ljql7uEhHELLIu_tgX1Cv_2Zwzu4dLJdD4cLqz1kcz74NgMvh12yoa078rOIgrfd3_U-F7T9u32x7ggiyXpAmRoi8FK6SevY4ArTDl4qYepjgyth_1ZOx44lEZvhtahhm9hHlGvrr0dxokcnhCJKjfFrbQrAFBCT9uDqzK4-VwXZi7TCCSWAUr6psAUaMtljVbpgQiXTne_tzJi0OUvVHiUJxJOyPsjChmibAk6Be6trE4kImt5ORjTEgZvr4G2-pCISgIwZG05hIdKYiE9dpXHqWLg_So1AZ5ybUA_SgkFZ_6QcpOWnDTB7YmnJlczWb2TnvG_4Oo8pRrZ59SASzZXHtKDfb0QswS9c","ims_user_id":7282103,"token_type":"Bearer","expires_in":3600,"expiration":1602182617,"refresh_token_expiration":1604771017,"scope":"ibm
```
Comment (agrare) [inline spec/vcr_cassettes/manageiq/providers/ibm_cloud/vpc/cloud_manager/refresher.yml:45]: So this is going to start failing 1 hour after this was recorded, I had been manually editing the expiration but this is obviously terrible :laughing: 

I opened a PR https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/81 so that this will be done automatically

Comment (blancett) [reply spec/vcr_cassettes/manageiq/providers/ibm_cloud/vpc/cloud_manager/refresher.yml:45]: Wow that is painful - ok will make sure to update till 81 is merged.

Comment (agrare) [reply spec/vcr_cassettes/manageiq/providers/ibm_cloud/vpc/cloud_manager/refresher.yml:45]: Okay @blancett go ahead and re-record this, should be good to go

**Review by agrare — APPROVED** [review]
:+1: LGTM

**Comment by agrare** [comment]
@blancett now you just need to rerun the spec locally to generate a new vcr_cassette

**Comment by blancett** [comment]
@agrare I recorded and pushed.

**Comment by agrare** [comment]
Hm looks like this pulled in some commits from master instead of rebasing

**Comment by blancett** [comment]
> Hm looks like this pulled in some commits from master instead of rebasing

**Comment by agrare** [comment]
@blancett looks like a bad rebase/merge

**Comment by agrare** [comment]
@blancett still looks like a bad rebase so hard to say if it is because of one of those commits or not
